// antd-mobile/cjs/components/progress-circle/progress-circle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProgressCircle = void 0;

var _react = _interopRequireDefault(require("react"));

var _rcProgress = require("rc-progress");

var _nativeProps = require("../../utils/native-props");

var _withDefaultProps = require("../../utils/with-default-props");

var _convertPx = require("../../utils/convert-px");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var classPrefix = "adm-progress";
var ProgressCircle = (0, _withDefaultProps.withDefaultProps)({
  percent: 0,
  strokeColor: '#1677FF'
})(function (props) {
  var strokeWidth = props.strokeWidth,
      size = props.size;

  if (strokeWidth === undefined) {
    strokeWidth = (0, _convertPx.convertPx)(6);
  }

  if (size === undefined) {
    size = (0, _convertPx.convertPx)(50);
  }

  var contentStyle = {
    width: size + "px",
    height: size + "px"
  };
  return (0, _nativeProps.withNativeProps)(props, /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-circle-wrap"
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-circle-content",
    style: contentStyle
  }, /*#__PURE__*/_react["default"].createElement(_rcProgress.Circle, {
    prefixCls: classPrefix,
    percent: props.percent,
    strokeWidth: strokeWidth,
    trailWidth: strokeWidth,
    trailColor: '#E5E5E5',
    strokeColor: props.strokeColor
  }), /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-circle-info"
  }, props.children))));
});
exports.ProgressCircle = ProgressCircle;
