// antd-mobile/es/components/search/search.js
import React, { useState, useRef, forwardRef, useImperativeHandle } from 'react';
import classNames from 'classnames';
import Input from '../input';
import { withNativeProps } from '../../utils/native-props';
import { mergeProps } from '../../utils/with-default-props';
import { SearchOutline } from 'antd-mobile-icons';
import { usePropsValue } from '../../utils/use-props-value';
import { useConfig } from '../config-provider';
var classPrefix = "adm-search";
var defaultProps = {
  clearable: true,
  showCancelButton: false,
  defaultValue: ''
};
export var Search = /*#__PURE__*/forwardRef(function (p, ref) {
  var _classNames;

  var _a;

  var props = mergeProps(defaultProps, p);

  var _usePropsValue = usePropsValue(props),
      value = _usePropsValue[0],
      setValue = _usePropsValue[1];

  var _useState = useState(false),
      hasFocus = _useState[0],
      setHasFocus = _useState[1];

  var inputRef = useRef(null);
  useImperativeHandle(ref, function () {
    return {
      clear: function clear() {
        var _a;

        return (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.clear();
      },
      focus: function focus() {
        var _a;

        return (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.focus();
      },
      blur: function blur() {
        var _a;

        return (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.blur();
      }
    };
  });

  var _useConfig = useConfig(),
      locale = _useConfig.locale;

  return withNativeProps(props, /*#__PURE__*/React.createElement("div", {
    className: classNames(classPrefix, (_classNames = {}, _classNames[classPrefix + "-active"] = hasFocus, _classNames))
  }, /*#__PURE__*/React.createElement("div", {
    className: classPrefix + "-input-box"
  }, /*#__PURE__*/React.createElement("div", {
    className: classPrefix + "-input-box-icon"
  }, /*#__PURE__*/React.createElement(SearchOutline, null)), /*#__PURE__*/React.createElement(Input, {
    ref: inputRef,
    className: classPrefix + "-input",
    value: value,
    onChange: setValue,
    maxLength: props.maxLength,
    placeholder: props.placeholder,
    clearable: props.clearable,
    onFocus: function onFocus(e) {
      var _a;

      setHasFocus(true);
      (_a = props.onFocus) === null || _a === void 0 ? void 0 : _a.call(props, e);
    },
    onBlur: function onBlur(e) {
      var _a;

      setHasFocus(false);
      (_a = props.onBlur) === null || _a === void 0 ? void 0 : _a.call(props, e);
    },
    onClear: props.onClear,
    type: 'search',
    onEnterPress: function onEnterPress() {
      var _a, _b;

      (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.blur();
      (_b = props.onSearch) === null || _b === void 0 ? void 0 : _b.call(props, value);
    }
  })), props.showCancelButton && hasFocus && /*#__PURE__*/React.createElement("div", {
    className: classPrefix + "-suffix"
  }, /*#__PURE__*/React.createElement("a", {
    onMouseDown: function onMouseDown(e) {
      e.preventDefault();
    },
    onTouchStart: function onTouchStart(e) {
      e.preventDefault();
    },
    onClick: function onClick() {
      var _a, _b, _c;

      (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.clear();
      (_b = inputRef.current) === null || _b === void 0 ? void 0 : _b.blur();
      (_c = props.onCancel) === null || _c === void 0 ? void 0 : _c.call(props);
    }
  }, (_a = props.cancelText) !== null && _a !== void 0 ? _a : locale.common.cancel))));
});
