// antd-mobile/2x/es/components/picker-view/column.js
import React, { useLayoutEffect, useRef, useState } from 'react';
import { useSpring, animated } from '@react-spring/web';
import { useDrag } from '@use-gesture/react';
import { convertPx } from '../../utils/convert-px';
import { rubberbandIfOutOfBounds } from '../../utils/rubberband';
import { bound } from '../../utils/bound';
import { useDebounceFn } from 'ahooks';
var classPrefix = "adm-picker-view";
export var Column = function Column(props) {
  var itemHeight = convertPx(34);
  var value = props.value,
      onSelect = props.onSelect,
      column = props.column;

  var _useSpring = useSpring(function () {
    return {
      from: {
        y: 0
      },
      config: {
        tension: 400,
        mass: 0.8
      }
    };
  }),
      y = _useSpring[0].y,
      api = _useSpring[1];

  var draggingRef = useRef(false);

  var _useState = useState({}),
      flag = _useState[0],
      setFlag = _useState[1];

  useLayoutEffect(function () {
    if (draggingRef.current) return;
    if (!value) return;
    var targetIndex = column.findIndex(function (item) {
      return item.value === value;
    });
    if (targetIndex < 0) return;
    var finalPosition = targetIndex * -itemHeight;
    api.start({
      y: finalPosition,
      immediate: y.idle
    });
  }, [value, column, flag]);
  useLayoutEffect(function () {
    if (column.length === 0) {
      if (value !== null) {
        props.onSelect(null);
      }
    } else {
      if (!column.some(function (item) {
        return item.value === value;
      })) {
        var firstItem = column[0];
        props.onSelect(firstItem.value);
      }
    }
  }, [column, value]);

  var _useDebounceFn = useDebounceFn(function () {
    setFlag({});
  }, {
    wait: 1000,
    leading: false,
    trailing: true
  }),
      debouncedUpdateFlag = _useDebounceFn.run;

  function scrollSelect(index) {
    var finalPosition = index * -itemHeight;
    api.start({
      y: finalPosition
    });
    onSelect(column[index].value);
    debouncedUpdateFlag();
  }

  var bind = useDrag(function (state) {
    draggingRef.current = true;
    var min = -((column.length - 1) * itemHeight);
    var max = 0;

    if (state.last) {
      draggingRef.current = false;
      var position = state.offset[1] + state.velocity[1] * state.direction[1] * 50;
      var targetIndex = -Math.round(bound(position, min, max) / itemHeight);
      scrollSelect(targetIndex);
    } else {
      var _position = state.offset[1];
      api.start({
        y: rubberbandIfOutOfBounds(_position, min, max, itemHeight * 50, 0.2)
      });
    }
  }, {
    axis: 'y',
    from: function from() {
      return [0, y.get()];
    },
    filterTaps: true
  });
  return /*#__PURE__*/React.createElement("div", Object.assign({
    className: classPrefix + "-column"
  }, bind()), /*#__PURE__*/React.createElement(animated.div, {
    style: {
      y: y
    },
    className: classPrefix + "-column-wheel"
  }, column.map(function (item, index) {
    function handleClick() {
      draggingRef.current = false;
      scrollSelect(index);
    }

    return /*#__PURE__*/React.createElement("div", {
      key: item.value,
      className: classPrefix + "-column-item",
      onClick: handleClick
    }, /*#__PURE__*/React.createElement("div", {
      className: classPrefix + "-column-item-label"
    }, item.label));
  })));
};
