// antd-mobile/2x/cjs/components/dialog/dialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Dialog = void 0;

var _react = _interopRequireDefault(require("react"));

var _withDefaultProps = require("../../utils/with-default-props");

var _classnames = _interopRequireDefault(require("classnames"));

var _mask = _interopRequireDefault(require("../mask"));

var _dialogActionButton = require("./dialog-action-button");

var _image = _interopRequireDefault(require("../image"));

var _space = _interopRequireDefault(require("../space"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var __awaiter = void 0 && (void 0).__awaiter || function (thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function (resolve) {
      resolve(value);
    });
  }

  return new (P || (P = Promise))(function (resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }

    function rejected(value) {
      try {
        step(generator["throw"](value));
      } catch (e) {
        reject(e);
      }
    }

    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }

    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
};

var classPrefix = "adm-dialog";
var defaultProps = {
  actions: [],
  closeOnAction: false,
  closeOnMaskClick: false
};

var Dialog = function Dialog(p) {
  var props = (0, _withDefaultProps.mergeProps)(defaultProps, p);
  return /*#__PURE__*/_react["default"].createElement(_mask["default"], {
    visible: props.visible,
    destroyOnClose: true,
    getContainer: props.getContainer,
    afterClose: props.afterClose,
    onMaskClick: props.closeOnMaskClick ? props.onClose : undefined,
    style: props.maskStyle,
    className: (0, _classnames["default"])(classPrefix + "-mask", props.maskClassName)
  }, /*#__PURE__*/_react["default"].createElement("div", {
    onClick: function onClick(e) {
      return e.stopPropagation();
    },
    className: classPrefix + "-wrap"
  }, !!props.image && /*#__PURE__*/_react["default"].createElement(_image["default"], {
    src: props.image,
    alt: 'dialog header image',
    width: '100%'
  }), /*#__PURE__*/_react["default"].createElement("div", {
    style: props.bodyStyle,
    className: (0, _classnames["default"])(classPrefix + "-body", props.bodyClassName)
  }, /*#__PURE__*/_react["default"].createElement(_space["default"], {
    direction: 'vertical',
    block: true
  }, !!props.header && /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-body-header-wrapper"
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-body-header"
  }, props.header)), !!props.title && /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-body-title"
  }, props.title), !!props.content && /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-body-message-wrapper"
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-body-message"
  }, props.content)))), /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-footer"
  }, props.actions.map(function (row, index) {
    var actions = Array.isArray(row) ? row : [row];
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: classPrefix + "-action-row",
      key: index
    }, actions.map(function (action, index) {
      return /*#__PURE__*/_react["default"].createElement(_dialogActionButton.DialogActionButton, {
        key: action.key,
        action: action,
        onAction: function onAction() {
          return __awaiter(void 0, void 0, void 0, /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
            var _a, _b;

            return regeneratorRuntime.wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    _context.next = 2;
                    return (_a = props.onAction) === null || _a === void 0 ? void 0 : _a.call(props, action, index);

                  case 2:
                    if (props.closeOnAction) {
                      (_b = props.onClose) === null || _b === void 0 ? void 0 : _b.call(props);
                    }

                  case 3:
                  case "end":
                    return _context.stop();
                }
              }
            }, _callee);
          }));
        }
      });
    }));
  }))));
};

exports.Dialog = Dialog;
