// antd-mobile/cjs/components/progress-bar/progress-bar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProgressBar = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _convertPx = require("../../utils/convert-px");

var _nativeProps = require("../../utils/native-props");

var _withDefaultProps = require("../../utils/with-default-props");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var classPrefix = "adm-progress-bar";
var ProgressBar = (0, _withDefaultProps.withDefaultProps)({
  percent: 0,
  strokeColor: '#1677FF'
})(function (props) {
  var strokeWidth = props.strokeWidth;

  if (strokeWidth === undefined) {
    strokeWidth = (0, _convertPx.convertPx)(3);
  }

  var trailStyle = {
    height: strokeWidth + "px",
    borderRadius: strokeWidth / 2 + "px",
    background: '#e5e5e5'
  };
  var pathStyle = {
    height: strokeWidth + "px",
    borderRadius: strokeWidth / 2 + "px",
    width: props.percent + "%",
    background: props.strokeColor,
    transition: 'width 0.3s'
  };
  return (0, _nativeProps.withNativeProps)(props, /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-trail",
    style: trailStyle
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])(classPrefix + "-path"),
    style: pathStyle
  }))));
});
exports.ProgressBar = ProgressBar;
