// antd-mobile/cjs/components/dropdown/dropdown.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _ahooks = require("ahooks");

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireWildcard(require("react"));

var _popup = _interopRequireDefault(require("../popup"));

var _item = _interopRequireWildcard(require("./item"));

var _nativeProps = require("../../utils/native-props");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var classPrefix = "adm-dropdown";

var Dropdown = function Dropdown(props) {
  var _classNames;

  var _useControllableValue = (0, _ahooks.useControllableValue)(props, {
    valuePropName: 'activeKey'
  }),
      value = _useControllableValue[0],
      onChange = _useControllableValue[1];

  var navRef = (0, _react.useRef)(null);
  var contentRef = (0, _react.useRef)(null); // 点击外部区域，关闭

  (0, _ahooks.useClickAway)(function () {
    onChange(undefined);
  }, [navRef, contentRef]); // 计算 navs 的 top 值

  var _useState = (0, _react.useState)(),
      top = _useState[0],
      setTop = _useState[1];

  var containerRef = (0, _react.useRef)(null);
  (0, _react.useEffect)(function () {
    if (value) {
      var rect = containerRef.current.getBoundingClientRect();
      setTop(rect.bottom);
    }
  }, [value]);

  var changeActive = function changeActive(key) {
    if (value === key) {
      onChange(undefined);
    } else {
      onChange(key);
    }
  };

  var popupForceRender = false;
  var items = [];

  var navs = _react["default"].Children.map(props.children, function (child) {
    if ( /*#__PURE__*/_react["default"].isValidElement(child)) {
      var childProps = Object.assign(Object.assign({}, child.props), {
        onClick: function onClick() {
          changeActive(child.key);
        },
        active: child.key === value
      });
      items.push(child);
      if (child.props.forceRender) popupForceRender = true;
      return /*#__PURE__*/(0, _react.cloneElement)(child, childProps);
    } else {
      return child;
    }
  });

  return (0, _nativeProps.withNativeProps)(props, /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])(classPrefix, (_classNames = {}, _classNames[classPrefix + "-open"] = !!value, _classNames)),
    ref: containerRef
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-nav",
    ref: navRef
  }, navs), /*#__PURE__*/_react["default"].createElement(_popup["default"], {
    visible: !!value,
    position: 'top',
    className: classPrefix + "-popup",
    maskClassName: classPrefix + "-popup-mask",
    bodyClassName: classPrefix + "-popup-body",
    style: {
      top: top
    },
    forceRender: popupForceRender
  }, /*#__PURE__*/_react["default"].createElement("div", {
    ref: contentRef
  }, items.map(function (item) {
    var isActive = item.key === value;
    return /*#__PURE__*/_react["default"].createElement(_item.ItemChildrenWrap, {
      key: item.key,
      active: isActive,
      forceRender: item.props.forceRender,
      destroyOnClose: item.props.destroyOnClose
    }, item.props.children);
  })))));
};

Dropdown.Item = _item["default"];
var _default = Dropdown;
exports["default"] = _default;
