// antd-mobile/2x/cjs/components/image/image.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Image = void 0;

var _withDefaultProps = require("../../utils/with-default-props");

var _react = _interopRequireWildcard(require("react"));

var _nativeProps = require("../../utils/native-props");

var _antdMobileIcons = require("antd-mobile-icons");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var classPrefix = "adm-image";
var defaultProps = {
  fit: 'fill',
  placeholder: /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-tip"
  }, /*#__PURE__*/_react["default"].createElement(_antdMobileIcons.PictureOutline, null)),
  fallback: /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-tip"
  }, /*#__PURE__*/_react["default"].createElement(_antdMobileIcons.PictureWrongOutline, null))
};

var Image = function Image(p) {
  var props = (0, _withDefaultProps.mergeProps)(defaultProps, p);

  var _useState = (0, _react.useState)(false),
      loaded = _useState[0],
      setLoaded = _useState[1];

  var _useState2 = (0, _react.useState)(false),
      failed = _useState2[0],
      setFailed = _useState2[1];

  function renderInner() {
    if (failed) {
      return props.fallback;
    }

    return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, !loaded && props.placeholder, /*#__PURE__*/_react["default"].createElement("img", {
      className: classPrefix + "-img",
      src: props.src,
      alt: props.alt,
      onClick: props.onClick,
      onLoad: function onLoad() {
        setLoaded(true);
      },
      onError: function onError(e) {
        var _a;

        setFailed(true);
        (_a = props.onError) === null || _a === void 0 ? void 0 : _a.call(props, e);
      },
      style: {
        objectFit: props.fit,
        display: loaded ? 'block' : 'none'
      },
      crossOrigin: props.crossOrigin,
      decoding: props.decoding,
      loading: props.loading,
      referrerPolicy: props.referrerPolicy,
      sizes: props.sizes,
      srcSet: props.srcSet,
      useMap: props.useMap
    }));
  }

  return (0, _nativeProps.withNativeProps)(props, /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix,
    style: {
      width: props.width,
      height: props.height
    }
  }, renderInner()));
};

exports.Image = Image;
