// antd-mobile/es/components/number-keyboard/number-keyboard.js
import React, { useRef, useMemo } from 'react';
import classNames from 'classnames';
import { DownOutline, TextDeletionOutline } from 'antd-mobile-icons';
import { mergeProps } from '../../utils/with-default-props';
import { shuffle } from '../../utils/shuffle';
import Popup from '../popup';
import { useClickAway } from 'ahooks';
import { usePropsValue } from '../../utils/use-props-value';
import { withNativeProps } from '../../utils/native-props';
var classPrefix = 'adm-number-keyboard';
var defaultProps = {
  defaultVisible: false,
  randomOrder: false,
  showCloseButton: true,
  confirmText: null,
  closeOnBlur: true,
  closeOnConfirm: true
};
export var NumberKeyboard = function NumberKeyboard(p) {
  var props = mergeProps(defaultProps, p);
  var title = props.title,
      getContainer = props.getContainer,
      confirmText = props.confirmText,
      customKey = props.customKey,
      randomOrder = props.randomOrder,
      showCloseButton = props.showCloseButton,
      onInput = props.onInput,
      onDelete = props.onDelete;

  var _usePropsValue = usePropsValue({
    value: props.visible,
    defaultValue: props.defaultVisible,
    onChange: function onChange(val) {
      var _a;

      if (!val) {
        (_a = props.onClose) === null || _a === void 0 ? void 0 : _a.call(props);
      }
    }
  }),
      visible = _usePropsValue[0],
      setVisible = _usePropsValue[1];

  var keyboardRef = useRef(null);
  useClickAway(function () {
    var _a;

    if (props.closeOnBlur && visible) {
      (_a = props.onBlur) === null || _a === void 0 ? void 0 : _a.call(props);
      setVisible(false);
    }
  }, keyboardRef);
  var keys = useMemo(function () {
    var defaultKeys = ['1', '2', '3', '4', '5', '6', '7', '8', '9'];
    var keyList = randomOrder ? shuffle(defaultKeys) : defaultKeys;
    keyList.push('0');

    if (confirmText) {
      keyList.push(customKey || '');
    } else {
      keyList.splice(9, 0, customKey || '');
      keyList.push('BACKSPACE');
    }

    return keyList;
  }, [customKey, confirmText, randomOrder, randomOrder && visible]); // 点击键盘按键

  var onKeyPress = function onKeyPress(key) {
    var _a;

    switch (key) {
      case 'BACKSPACE':
        onDelete === null || onDelete === void 0 ? void 0 : onDelete();
        break;

      case 'OK':
        (_a = props.onConfirm) === null || _a === void 0 ? void 0 : _a.call(props);

        if (props.closeOnConfirm) {
          setVisible(false);
        }

        break;

      default:
        onInput === null || onInput === void 0 ? void 0 : onInput(key);
        break;
    }
  }; // 渲染 title 和 close button


  var renderHeader = function renderHeader() {
    if (!showCloseButton && !title) return null;
    return /*#__PURE__*/React.createElement("div", {
      className: classNames(classPrefix + "-header", {
        'with-title': !!title
      })
    }, title && /*#__PURE__*/React.createElement("div", {
      className: classPrefix + "-title"
    }, title), showCloseButton && /*#__PURE__*/React.createElement("span", {
      className: classPrefix + "-header-close-button",
      onClick: function onClick() {
        setVisible(false);
      },
      role: 'button',
      title: 'CLOSE'
    }, /*#__PURE__*/React.createElement(DownOutline, null)));
  }; // 渲染基础键盘按键


  var renderKey = function renderKey(key, index) {
    var isNumberKey = /^\d$/.test(key);
    var className = classNames(classPrefix + "-key", {
      'number-key': isNumberKey,
      'sign-key': !isNumberKey && key,
      'mid-key': index === 9 && !!confirmText
    });
    return /*#__PURE__*/React.createElement("div", {
      key: key,
      className: className,
      onClick: function onClick() {
        return key && onKeyPress(key);
      },
      title: key,
      role: 'button'
    }, key === 'BACKSPACE' ? /*#__PURE__*/React.createElement(TextDeletionOutline, null) : key);
  };

  return /*#__PURE__*/React.createElement(Popup, {
    visible: visible,
    getContainer: getContainer,
    mask: false,
    afterClose: props.afterClose,
    afterShow: props.afterShow,
    className: classPrefix + "-popup",
    stopPropagation: props.stopPropagation
  }, withNativeProps(props, /*#__PURE__*/React.createElement("div", {
    ref: keyboardRef,
    className: classPrefix
  }, renderHeader(), /*#__PURE__*/React.createElement("div", {
    className: classPrefix + "-wrapper"
  }, /*#__PURE__*/React.createElement("div", {
    className: classNames(classPrefix + "-main", {
      'confirmed-style': !!confirmText
    })
  }, keys.map(renderKey)), !!confirmText && /*#__PURE__*/React.createElement("div", {
    className: classPrefix + "-confirm"
  }, /*#__PURE__*/React.createElement("div", {
    className: classPrefix + "-key extra-key bs-key",
    onClick: function onClick() {
      return onKeyPress('BACKSPACE');
    },
    title: 'BACKSPACE',
    role: 'button'
  }, /*#__PURE__*/React.createElement(TextDeletionOutline, null)), /*#__PURE__*/React.createElement("div", {
    className: classPrefix + "-key extra-key ok-key",
    onClick: function onClick() {
      return onKeyPress('OK');
    },
    role: 'button'
  }, confirmText))))));
};
