// antd-mobile/es/components/progress-circle/progress-circle.js
import React from 'react';
import { withNativeProps } from '../../utils/native-props';
import { withDefaultProps } from '../../utils/with-default-props';
var classPrefix = "adm-progress-circle";
export var ProgressCircle = withDefaultProps({
  percent: 0,
  strokeColor: '#1677FF'
})(function (props) {
  var style = {
    '--percent': props.percent.toString()
  };
  return withNativeProps(props, /*#__PURE__*/React.createElement("div", {
    className: "" + classPrefix,
    style: style
  }, /*#__PURE__*/React.createElement("div", {
    className: classPrefix + "-content"
  }, /*#__PURE__*/React.createElement("svg", {
    className: classPrefix + "-svg"
  }, /*#__PURE__*/React.createElement("circle", {
    className: classPrefix + "-track",
    fill: 'transparent'
  }), /*#__PURE__*/React.createElement("circle", {
    className: classPrefix + "-fill",
    fill: 'transparent'
  })), /*#__PURE__*/React.createElement("div", {
    className: classPrefix + "-info"
  }, props.children))));
});
