// antd-mobile/cjs/components/desense-text/desense-text.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DesenseText = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _nativeProps = require("../../utils/native-props");

var _withDefaultProps = require("../../utils/with-default-props");

var _antdMobileIcons = require("antd-mobile-icons");

var _usePropsValue2 = require("../../utils/use-props-value");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var classPrefix = 'adm-desense-text';
var defaultProps = {
  defaultDesense: true
};

var DesenseText = function DesenseText(p) {
  var props = (0, _withDefaultProps.mergeProps)(defaultProps, p);
  var text = props.text,
      desenseText = props.desenseText;

  var _usePropsValue = (0, _usePropsValue2.usePropsValue)({
    value: props.desense,
    defaultValue: props.defaultDesense,
    onChange: props.onChange
  }),
      isDesense = _usePropsValue[0],
      setIsDesense = _usePropsValue[1];

  return (0, _nativeProps.withNativeProps)(props, /*#__PURE__*/_react["default"].createElement("span", {
    className: classPrefix
  }, isDesense ? desenseText : text, /*#__PURE__*/_react["default"].createElement("a", {
    className: (0, _classnames["default"])(classPrefix + "-icon-wrap", 'adm-plain-anchor'),
    onClick: function onClick() {
      setIsDesense(!isDesense);
    }
  }, isDesense ? /*#__PURE__*/_react["default"].createElement(_antdMobileIcons.EyeInvisibleOutline, null) : /*#__PURE__*/_react["default"].createElement(_antdMobileIcons.EyeOutline, null))));
};

exports.DesenseText = DesenseText;
