// antd-mobile/es/components/index-bar/index-bar.js
import React, { forwardRef, useRef, useState, useEffect, useImperativeHandle } from 'react';
import classNames from 'classnames';
import { withNativeProps } from '../../utils/native-props';
import { useThrottleFn } from 'ahooks';
import { mergeProps } from '../../utils/with-default-props';
import { Sidebar } from './sidebar';
import { IndexBarContext } from './context';
import { convertPx } from '../../utils/convert-px';
var classPrefix = "adm-index-bar";
var defaultProps = {
  sticky: true
};
export var IndexBar = /*#__PURE__*/forwardRef(function (p, ref) {
  var _classNames;

  var props = mergeProps(defaultProps, p);
  var titleHeight = convertPx(35);
  var bodyRef = useRef(null);

  var _useState = useState([]),
      indexes = _useState[0],
      setIndexes = _useState[1];

  var _useState2 = useState(indexes[0]),
      activeIndex = _useState2[0],
      setActiveIndex = _useState2[1];

  useImperativeHandle(ref, function () {
    return {
      scrollTo: scrollTo
    };
  });

  function scrollTo(index) {
    var body = bodyRef.current;
    if (!body) return;
    var children = body.children;

    for (var i = 0; i < children.length; i++) {
      var panel = children.item(i);
      if (!panel) continue;
      var panelIndex = panel.dataset['index'];

      if (panelIndex === index) {
        body.scrollTop = panel.offsetTop;
        setActiveIndex(index);
        return;
      }
    }
  }

  var _useThrottleFn = useThrottleFn(function () {
    var body = bodyRef.current;
    if (!body) return;
    var scrollTop = body.scrollTop;
    var children = body.getElementsByClassName(classPrefix + "-anchor");

    for (var i = 0; i < children.length; i++) {
      var panel = children.item(i);
      if (!panel) continue;
      var panelIndex = panel.dataset['index'];
      if (!panelIndex) continue;

      if (panel.offsetTop + panel.clientHeight - titleHeight > scrollTop) {
        setActiveIndex(panelIndex);
        return;
      }
    }
  }, {
    wait: 50,
    trailing: true,
    leading: true
  }),
      checkActiveIndex = _useThrottleFn.run;

  useEffect(function () {
    if (!indexes.length) return;
    checkActiveIndex();
  }, [indexes]);
  var element = withNativeProps(props, /*#__PURE__*/React.createElement("div", {
    className: classNames("" + classPrefix, (_classNames = {}, _classNames[classPrefix + "-sticky"] = props.sticky, _classNames))
  }, /*#__PURE__*/React.createElement(Sidebar, {
    indexes: indexes,
    activeIndex: activeIndex,
    onActive: function onActive(index) {
      scrollTo(index);
    }
  }), /*#__PURE__*/React.createElement("div", {
    className: classPrefix + "-body",
    ref: bodyRef,
    onScroll: checkActiveIndex
  }, props.children)));
  return /*#__PURE__*/React.createElement(IndexBarContext.Provider, {
    value: {
      indexes: indexes,
      setIndexes: setIndexes
    }
  }, element);
});
