// antd-mobile/cjs/components/tabs/tabs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tabs = exports.TabPane = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _nativeProps = require("../../utils/native-props");

var _useControllableValue = require("../../utils/use-controllable-value");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var classPrefix = "adm-tabs";

var TabPane = function TabPane() {
  return null;
};

exports.TabPane = TabPane;

var Tabs = function Tabs(props) {
  var _a;

  var childrenRecord = {};
  var firstActiveKey = null;
  var panes = [];

  _react["default"].Children.forEach(props.children, function (child, index) {
    if (! /*#__PURE__*/_react["default"].isValidElement(child)) return;
    var key = child.key;
    if (typeof key !== 'string') return;

    if (index === 0) {
      firstActiveKey = key;
    }

    childrenRecord[key] = child.props.children;
    panes.push(child);
  });

  var _useNewControllableVa = (0, _useControllableValue.useNewControllableValue)({
    value: props.activeKey,
    defaultValue: (_a = props.defaultActiveKey) !== null && _a !== void 0 ? _a : firstActiveKey,
    onChange: props.onChange
  }),
      activeKey = _useNewControllableVa[0],
      setActiveKey = _useNewControllableVa[1];

  return (0, _nativeProps.withNativeProps)(props, /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-tab-list"
  }, panes.map(function (pane) {
    var _classNames;

    return (0, _nativeProps.withNativeProps)(pane.props, /*#__PURE__*/_react["default"].createElement("div", {
      key: pane.key,
      className: classPrefix + "-tab-wrapper"
    }, /*#__PURE__*/_react["default"].createElement("div", {
      onClick: function onClick() {
        var key = pane.key;

        if (key === undefined || key === null) {
          return;
        }

        setActiveKey(key.toString());
      },
      className: (0, _classnames["default"])(classPrefix + "-tab", (_classNames = {}, _classNames[classPrefix + "-tab-active"] = pane.key === activeKey, _classNames))
    }, pane.props.title)));
  })), panes.map(function (pane) {
    if (pane.props.children === undefined) {
      return null;
    }

    if (pane.key === activeKey) {
      return /*#__PURE__*/_react["default"].createElement("div", {
        key: pane.key,
        className: classPrefix + "-content"
      }, pane.props.children);
    }

    if (pane.props.forceRender) {
      return /*#__PURE__*/_react["default"].createElement("div", {
        key: pane.key,
        style: {
          display: 'none'
        }
      }, pane.props.children);
    }

    return null;
  })));
};

exports.Tabs = Tabs;
