// antd-mobile/cjs/components/index-bar/index-bar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IndexBar = void 0;

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _nativeProps = require("../../utils/native-props");

var _ahooks = require("ahooks");

var _withDefaultProps = require("../../utils/with-default-props");

var _sidebar = require("./sidebar");

var _context = require("./context");

var _convertPx = require("../../utils/convert-px");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var classPrefix = "adm-index-bar";
var defaultProps = {
  sticky: true
};
var IndexBar = /*#__PURE__*/(0, _react.forwardRef)(function (p, ref) {
  var _classNames;

  var props = (0, _withDefaultProps.mergeProps)(defaultProps, p);
  var titleHeight = (0, _convertPx.convertPx)(35);
  var bodyRef = (0, _react.useRef)(null);

  var _useState = (0, _react.useState)([]),
      indexes = _useState[0],
      setIndexes = _useState[1];

  var _useState2 = (0, _react.useState)(indexes[0]),
      activeIndex = _useState2[0],
      setActiveIndex = _useState2[1];

  (0, _react.useImperativeHandle)(ref, function () {
    return {
      scrollTo: scrollTo
    };
  });

  function scrollTo(index) {
    var body = bodyRef.current;
    if (!body) return;
    var children = body.children;

    for (var i = 0; i < children.length; i++) {
      var panel = children.item(i);
      if (!panel) continue;
      var panelIndex = panel.dataset['index'];

      if (panelIndex === index) {
        body.scrollTop = panel.offsetTop;
        setActiveIndex(index);
        return;
      }
    }
  }

  var _useThrottleFn = (0, _ahooks.useThrottleFn)(function () {
    var body = bodyRef.current;
    if (!body) return;
    var scrollTop = body.scrollTop;
    var children = body.getElementsByClassName(classPrefix + "-anchor");

    for (var i = 0; i < children.length; i++) {
      var panel = children.item(i);
      if (!panel) continue;
      var panelIndex = panel.dataset['index'];
      if (!panelIndex) continue;

      if (panel.offsetTop + panel.clientHeight - titleHeight > scrollTop) {
        setActiveIndex(panelIndex);
        return;
      }
    }
  }, {
    wait: 50,
    trailing: true,
    leading: true
  }),
      checkActiveIndex = _useThrottleFn.run;

  (0, _react.useEffect)(function () {
    if (!indexes.length) return;
    checkActiveIndex();
  }, [indexes]);
  var element = (0, _nativeProps.withNativeProps)(props, /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])("" + classPrefix, (_classNames = {}, _classNames[classPrefix + "-sticky"] = props.sticky, _classNames))
  }, /*#__PURE__*/_react["default"].createElement(_sidebar.Sidebar, {
    indexes: indexes,
    activeIndex: activeIndex,
    onActive: function onActive(index) {
      scrollTo(index);
    }
  }), /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-body",
    ref: bodyRef,
    onScroll: checkActiveIndex
  }, props.children)));
  return /*#__PURE__*/_react["default"].createElement(_context.IndexBarContext.Provider, {
    value: {
      indexes: indexes,
      setIndexes: setIndexes
    }
  }, element);
});
exports.IndexBar = IndexBar;
