// antd-mobile/2x/cjs/components/check-list/check-list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CheckList = void 0;

var _react = _interopRequireDefault(require("react"));

var _nativeProps = require("../../utils/native-props");

var _list = _interopRequireDefault(require("../list"));

var _withDefaultProps = require("../../utils/with-default-props");

var _context = require("./context");

var _useControllableValue = require("../../utils/use-controllable-value");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var defaultProps = {
  multiple: false,
  defaultValue: []
};

var CheckList = function CheckList(p) {
  var props = (0, _withDefaultProps.mergeProps)(defaultProps, p);

  var _useNewControllableVa = (0, _useControllableValue.useNewControllableValue)(props),
      value = _useNewControllableVa[0],
      setValue = _useNewControllableVa[1];

  function check(val) {
    if (props.multiple) {
      setValue([].concat(value, [val]));
    } else {
      setValue([val]);
    }
  }

  function uncheck(val) {
    setValue(value.filter(function (item) {
      return item !== val;
    }));
  }

  return (0, _nativeProps.withNativeProps)(props, /*#__PURE__*/_react["default"].createElement(_context.CheckListContext.Provider, {
    value: {
      value: value,
      check: check,
      uncheck: uncheck
    }
  }, /*#__PURE__*/_react["default"].createElement(_list["default"], {
    mode: props.mode
  }, props.children)));
};

exports.CheckList = CheckList;
