// antd-mobile/2x/es/components/rate/rate.js
import React from 'react';
import classNames from 'classnames';
import { withNativeProps } from '../../utils/native-props';
import { withDefaultProps } from '../../utils/with-default-props';
import { StarFill } from 'antd-mobile-icons';
import { useNewControllableValue } from '../../utils/use-controllable-value';
var classPrefix = "adm-rate";
var defaultProps = {
  count: 5,
  allowHalf: false,
  character: /*#__PURE__*/React.createElement(StarFill, null),
  defaultValue: 0,
  readonly: false,
  allowClear: true
};
export var Rate = withDefaultProps(defaultProps)(function (props) {
  var _useNewControllableVa = useNewControllableValue(props),
      value = _useNewControllableVa[0],
      setValue = _useNewControllableVa[1];

  var starList = Array(props.count).fill(null);

  function renderStar(v, half) {
    var _classNames;

    return /*#__PURE__*/React.createElement("div", {
      className: classNames(classPrefix + "-star", (_classNames = {}, _classNames[classPrefix + "-star-active"] = value >= v, _classNames[classPrefix + "-star-half"] = half, _classNames[classPrefix + "-star-readonly"] = props.readonly, _classNames)),
      onClick: function onClick() {
        if (props.readonly) return;

        if (props.allowClear && value === v) {
          setValue(0);
        } else {
          setValue(v);
        }
      }
    }, props.character);
  }

  return withNativeProps(props, /*#__PURE__*/React.createElement("div", {
    className: classPrefix
  }, starList.map(function (_, i) {
    return /*#__PURE__*/React.createElement("div", {
      key: i,
      className: classNames(classPrefix + "-box")
    }, props.allowHalf && renderStar(i + 0.5, true), renderStar(i + 1, false));
  })));
});
