// antd-mobile/2x/cjs/components/checkbox/group.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Group = void 0;

var _react = _interopRequireDefault(require("react"));

var _withDefaultProps = require("../../utils/with-default-props");

var _groupContext = require("./group-context");

var _useControllableValue = require("../../utils/use-controllable-value");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var defaultProps = {
  disabled: false,
  defaultValue: []
};
var Group = (0, _withDefaultProps.withDefaultProps)(defaultProps)(function (props) {
  var _useNewControllableVa = (0, _useControllableValue.useNewControllableValue)(props),
      value = _useNewControllableVa[0],
      setValue = _useNewControllableVa[1];

  return /*#__PURE__*/_react["default"].createElement(_groupContext.CheckboxGroupContext.Provider // TODO: 性能优化
  , {
    // TODO: 性能优化
    value: {
      value: value,
      disabled: props.disabled,
      check: function check(v) {
        setValue([].concat(value, [v]));
      },
      uncheck: function uncheck(v) {
        setValue(value.filter(function (item) {
          return item !== v;
        }));
      }
    }
  }, props.children);
});
exports.Group = Group;
