// antd-mobile/cjs/components/grid/grid.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridItem = exports.Grid = void 0;

var _withDefaultProps = require("../../utils/with-default-props");

var _react = _interopRequireDefault(require("react"));

var _nativeProps = require("../../utils/native-props");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var classPrefix = "adm-grid";
var defaultProps = {
  gap: 0
};
var Grid = (0, _withDefaultProps.withDefaultProps)(defaultProps)(function (props) {
  var gapStyle = {};
  var gap = props.gap;

  if (gap) {
    var _ref = Array.isArray(gap) ? gap : [gap, gap],
        horizontalGap = _ref[0],
        verticalGap = _ref[1];

    gapStyle = Object.assign(Object.assign({}, gapStyle), {
      '--vertical-gap': typeof verticalGap === 'number' ? verticalGap + "px" : verticalGap,
      '--horizontal-gap': typeof horizontalGap === 'number' ? horizontalGap + "px" : horizontalGap
    });
  }

  return (0, _nativeProps.withNativeProps)(props, /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix,
    style: Object.assign(Object.assign({}, gapStyle), {
      '--columns': props.columns
    })
  }, props.children));
});
exports.Grid = Grid;
var GridItem = (0, _withDefaultProps.withDefaultProps)({
  span: 1
})(function (props) {
  var itemStyle = {
    '--item-span': props.span
  };
  return (0, _nativeProps.withNativeProps)(props, /*#__PURE__*/_react["default"].createElement("div", {
    className: classPrefix + "-item",
    style: itemStyle
  }, props.children));
});
exports.GridItem = GridItem;
