// antd-mobile/cjs/components/swiper/swiper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Swiper = void 0;

var _react = _interopRequireWildcard(require("react"));

var _nativeProps = require("../../utils/native-props");

var _withDefaultProps = require("../../utils/with-default-props");

var _classnames = _interopRequireDefault(require("classnames"));

var _swiperItem = require("./swiper-item");

var _devLog = require("../../utils/dev-log");

var _web = require("@react-spring/web");

var _react2 = require("@use-gesture/react");

var _pageIndicator = _interopRequireDefault(require("../page-indicator"));

var _stagedComponents = require("staged-components");

var _useRefState2 = require("../../utils/use-ref-state");

var _bound = require("../../utils/bound");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var defaultProps = {
  defaultIndex: 0,
  allowTouchMove: true,
  autoplay: false,
  autoplayInterval: 3000,
  loop: true
};
var Swiper = /*#__PURE__*/(0, _react.forwardRef)((0, _stagedComponents.staged)(function (p, ref) {
  var props = (0, _withDefaultProps.mergeProps)(defaultProps, p);

  var _useMemo = (0, _react.useMemo)(function () {
    var count = 0;

    var validChildren = _react["default"].Children.map(props.children, function (child) {
      if (! /*#__PURE__*/_react["default"].isValidElement(child)) return null;

      if (child.type !== _swiperItem.SwiperItem) {
        (0, _devLog.devWarning)('Swiper', 'The children of `Swiper` must be `Swiper.Item` components.');
        return null;
      }

      count++;
      return child;
    });

    return {
      validChildren: validChildren,
      count: count
    };
  }, [props.children]),
      validChildren = _useMemo.validChildren,
      count = _useMemo.count;

  if (count === 0) {
    (0, _devLog.devWarning)('Swiper', '`Swiper` needs at least one child.');
    return null;
  }

  return function () {
    var loop = props.loop;

    if (count === 1 && loop) {
      (0, _devLog.devWarning)('Swiper', '`Swiper` needs at least two children to enable loop.');
      loop = false;
    }

    var trackRef = (0, _react.useRef)(null);

    function getWidth() {
      var track = trackRef.current;
      if (!track) return 0;
      return track.offsetWidth;
    }

    var _useState = (0, _react.useState)(props.defaultIndex),
        current = _useState[0],
        setCurrent = _useState[1];

    var _useRefState = (0, _useRefState2.useRefState)(false),
        dragging = _useRefState[0],
        setDragging = _useRefState[1],
        draggingRef = _useRefState[2];

    var _useSpring = (0, _web.useSpring)(function () {
      return {
        x: (0, _bound.bound)(current, 0, count - 1) * 100,
        config: {
          tension: 200,
          friction: 30
        },
        onRest: function onRest() {
          if (draggingRef.current) return;
          var rawX = x.get();
          var totalWidth = 100 * count;
          var standardX = modulus(rawX, totalWidth);
          if (standardX === rawX) return;
          api.start({
            x: standardX,
            immediate: true
          });
        }
      };
    }, [count]),
        x = _useSpring[0].x,
        api = _useSpring[1];

    var bind = (0, _react2.useDrag)(function (state) {
      var width = getWidth();
      if (!width) return;
      var _state$offset = state.offset,
          offsetX = _state$offset[0];
      setDragging(true);

      if (!state.last) {
        api.start({
          x: offsetX * 100 / width,
          immediate: true
        });
      } else {
        var index = Math.round((offsetX + Math.min(state.velocity[0] * 2000, width) * state.direction[0]) / width);
        swipeTo(index);
        window.setTimeout(function () {
          setDragging(false);
        });
      }
    }, {
      transform: function transform(_ref) {
        var x = _ref[0],
            y = _ref[1];
        return [-x, y];
      },
      from: function from() {
        var width = getWidth();
        return [x.get() / 100 * width, 0];
      },
      bounds: function bounds() {
        if (loop) return {};
        var width = getWidth();
        return {
          left: 0,
          right: (count - 1) * width
        };
      },
      rubberband: true,
      axis: 'x',
      preventScroll: true,
      pointer: {
        touch: true
      }
    });

    function swipeTo(index) {
      var _a, _b;

      if (loop) {
        var i = modulus(index, count);
        setCurrent(i);
        (_a = props.onIndexChange) === null || _a === void 0 ? void 0 : _a.call(props, i);
        api.start({
          x: index * 100
        });
      } else {
        var _i = (0, _bound.bound)(index, 0, count - 1);

        setCurrent(_i);
        (_b = props.onIndexChange) === null || _b === void 0 ? void 0 : _b.call(props, _i);
        api.start({
          x: _i * 100
        });
      }
    }

    function swipeNext() {
      swipeTo(Math.round(x.get() / 100) + 1);
    }

    function swipePrev() {
      swipeTo(Math.round(x.get() / 100) - 1);
    }

    (0, _react.useImperativeHandle)(ref, function () {
      return {
        swipeTo: swipeTo,
        swipeNext: swipeNext,
        swipePrev: swipePrev
      };
    });
    var autoplay = props.autoplay,
        autoplayInterval = props.autoplayInterval;
    (0, _react.useEffect)(function () {
      if (!autoplay || dragging) return;
      var interval = window.setInterval(function () {
        swipeNext();
      }, autoplayInterval);
      return function () {
        window.clearInterval(interval);
      };
    }, [autoplay, autoplayInterval, dragging]);
    return (0, _nativeProps.withNativeProps)(props, /*#__PURE__*/_react["default"].createElement("div", {
      className: 'adm-swiper'
    }, /*#__PURE__*/_react["default"].createElement("div", Object.assign({
      className: (0, _classnames["default"])('adm-swiper-track', {
        'adm-swiper-track-allow-touch-move': props.allowTouchMove
      }),
      onClickCapture: function onClickCapture(e) {
        if (draggingRef.current) {
          e.stopPropagation();
        }
      }
    }, props.allowTouchMove ? bind() : {}), /*#__PURE__*/_react["default"].createElement("div", {
      className: 'adm-swiper-track-inner',
      ref: trackRef
    }, _react["default"].Children.map(validChildren, function (child, index) {
      return /*#__PURE__*/_react["default"].createElement(_web.animated.div, {
        className: 'adm-swiper-slide',
        style: {
          x: x.to(function (x) {
            var position = -x + index * 100;

            if (loop) {
              var totalWidth = count * 100;
              var flagWidth = totalWidth / 2 - 10;
              position = modulus(position + flagWidth, totalWidth) - flagWidth;
            }

            return position + "%";
          }),
          left: "-" + index * 100 + "%"
        }
      }, child);
    }))), props.indicator === undefined ? /*#__PURE__*/_react["default"].createElement("div", {
      className: 'adm-swiper-indicator'
    }, /*#__PURE__*/_react["default"].createElement(_pageIndicator["default"], Object.assign({}, props.indicatorProps, {
      total: count,
      current: current
    }))) : props.indicator(count, current)));
  };
}));
exports.Swiper = Swiper;

function modulus(value, division) {
  var remainder = value % division;
  return remainder < 0 ? remainder + division : remainder;
}
