// antd-mobile/2x/es/components/check-list/check-list.js
import React from 'react';
import { withNativeProps } from '../../utils/native-props';
import List from '../list';
import { mergeProps } from '../../utils/with-default-props';
import { CheckListContext } from './context';
import { useNewControllableValue } from '../../utils/use-controllable-value';
var defaultProps = {
  multiple: false,
  defaultValue: []
};
export var CheckList = function CheckList(p) {
  var props = mergeProps(defaultProps, p);

  var _useNewControllableVa = useNewControllableValue(props),
      value = _useNewControllableVa[0],
      setValue = _useNewControllableVa[1];

  function check(val) {
    if (props.multiple) {
      setValue([].concat(value, [val]));
    } else {
      setValue([val]);
    }
  }

  function uncheck(val) {
    setValue(value.filter(function (item) {
      return item !== val;
    }));
  }

  return withNativeProps(props, /*#__PURE__*/React.createElement(CheckListContext.Provider, {
    value: {
      value: value,
      check: check,
      uncheck: uncheck
    }
  }, /*#__PURE__*/React.createElement(List, {
    mode: props.mode
  }, props.children)));
};
