// antd-mobile/2x/es/components/check-list/check-list-item.js
import React, { useContext } from 'react';
import List from '../list';
import { withNativeProps } from '../../utils/native-props';
import { CheckListContext } from './context';
import { devWarning } from '../../utils/dev-log';
import { CheckOutline } from 'antd-mobile-icons';
export var CheckListItem = function CheckListItem(props) {
  var context = useContext(CheckListContext);

  if (context === null) {
    devWarning('CheckList.Item', 'CheckList.Item can only be used inside CheckList.');
    return null;
  }

  var active = context.value.includes(props.value);
  var extra = /*#__PURE__*/React.createElement("div", {
    className: 'adm-check-list-item-extra'
  }, active ? /*#__PURE__*/React.createElement(CheckOutline, null) : null);
  return withNativeProps(props, /*#__PURE__*/React.createElement(List.Item, {
    title: props.title,
    description: props.description,
    prefix: props.prefix,
    onClick: function onClick(e) {
      var _a;

      if (active) {
        context.uncheck(props.value);
      } else {
        context.check(props.value);
      }

      (_a = props.onClick) === null || _a === void 0 ? void 0 : _a.call(props, e);
    },
    arrow: false,
    clickable: true,
    extra: extra
  }, props.children));
};
